### PR TITLE
Aggregated Logbook 

### PR DESCRIPTION
**Pull request by gardaholm**
as discussed in https://github.com/lemeryfertitta/Climbdex/issues/55

slightly updated version of `logbook_entries()` functions from @PTschaikner to create an aggregated logbook `logbook_entries_agg()` with all needed informations which are used to get personal stats about the each climb ('climb_uuid', 'climb_name', 'board', 'date', 'is_mirror', 'angle', 'tries', 'sessions' 'time_since').

instead of the login credentials (user_id and password) `logbook_entries_agg()` uses user_id and token to fetch the results.

### CODE REVIEW

File: src/boardlib/api/aurora.py [hunk 1]
```diff
@@ -580,5 +581,115 @@ def logbook_entries(board, username, password, grade_type="font", db_path=None):
     
     return full_logbook_df
 
+def bids_logbook_entries_agg(board, token, user_id, db_path=None):
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:584]: My suggestion in the other PR was that we could refactor that PR to adjust the params to get the output needed for the Climbdex changes. It looks like this PR just duplicates everything. E.g. we could refactor bids_logbook_entries to take a token instead of user/password and we can parametrize which entries are displayed in the output.

Seems like there is a lot of repeated code here that will be difficult to maintain in the future, and it's difficult to understand why it is this way and what might break if we made changes.

Let me know if I'm missing something here and there is a reason why both features can't use the same underlying functions with slightly different parameters

**Review by lemeryfertitta — CHANGES_REQUESTED** [review]

**Comment by gardaholm** [comment]
We refactored the function so the command `boardlib logbook <board_name> --username=<username> --output=<output_file_name>.csv --grade-type="hueco" --database=<local_database_file>` works like before (with username, password) and exports the logbook as csv. 

To get the aggregated version of the logbook, which is used to calculate bids/attempts we use the same function `logbook_entries()` in `get_bids()`, but with the argument aggregate=true, see [here](https://github.com/lemeryfertitta/Climbdex/blob/1a35c2b5d9215d474cd526a4b2170189bae7085d/climbdex/views.py#L158)
when aggregate is set to true then the `logbook_entries()` works with user_id and token from the cookie instead of the username and password from the terminal commands.

The refactor was a bit tricky, but all our tests worked.
In our climbdex+boardlib version we could tick a climb via the new menu item, and exporting the logbook via commandline showed the new entry in the csv file.

File: src/boardlib/api/aurora.py (hunk 1 above)
Comment (gardaholm) [reply src/boardlib/api/aurora.py]: makes sense, we tried our best to refactor the function, see comment below.

File: src/boardlib/__main__.py [hunk 2]
```diff
@@ -1,17 +1,16 @@
 import argparse
```
Comment (lemeryfertitta) [inline src/boardlib/__main__.py:1]: There are a lot of changes in this file, but it seems to me like at most this should just be a one-line change to add `aggregate=False`, but really there shouldn't be any changes at all because aggregate is already defaulting to False. Can we revert this file please? There are multiple regressions, e.g. to the command output, the imports, and the styling.

File: src/boardlib/api/aurora.py [hunk 3]
```diff
@@ -367,31 +365,42 @@ def get_bids_logbook(board, token, user_id):
...
-            "comment": raw_entry["comment"],
-            "climbed_at": raw_entry["climbed_at"],
-            "created_at": raw_entry["created_at"],
-        }
 
+        if aggregate:
+
+            entry = {
+                "climb_uuid": raw_entry["climb_uuid"],
+                "board": board,
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:381]: Why does board need to be in the entry? It's already passed as an argument, so the caller already knows what the board is.

File: src/boardlib/api/aurora.py [hunk 4]
```diff
@@ -367,31 +365,42 @@ def get_bids_logbook(board, token, user_id):
...
-            "climb_name": climb_name,
-            "angle": raw_entry["angle"],
-            "is_mirror": raw_entry["is_mirror"],
-            "bid_count": raw_entry["bid_count"],
-            "comment": raw_entry["comment"],
-            "climbed_at": raw_entry["climbed_at"],
-            "created_at": raw_entry["created_at"],
-        }
 
+        if aggregate:
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:377]: It doesn't seem like we really need to add this if clause at all. It's such a minor performance gain to drop a few fields from the `aggregate` version and perhaps we'd eventually want all of these fields in both versions anyways (for example, comment seems like something we'd eventually display).

File: src/boardlib/api/aurora.py [hunk 5]
```diff
@@ -541,24 +591,31 @@ def calculate_tries_total(group):
     return group
 
 
-def logbook_entries(board, username, password, grade_type="font", db_path=None):
-    login_info = login(board, username, password)
-    token = login_info["token"]
-    user_id = login_info["user_id"]
-    
-    bids_entries = list(bids_logbook_entries(board, username, password, db_path))
+def logbook_entries(board, username, password=None, token=None, user_id=None, grade_type="font", db_path=None, aggregate=False):
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:594]: We don't need a function to support both forms of arguments. We should just have the command line call an extra function to fetch the token and then have `logbook_entries` require a token.

File: src/boardlib/api/aurora.py [hunk 6]
```diff
@@ -473,56 +492,87 @@ def combine_ascents_and_bids(ascents_df, bids_summary, db_path, grades_dict, gra
...
+                }
+            final_logbook.append(entry)
 
     for _, bid_row in bids_summary.iterrows():
-        if db_path:
-            difficulty = get_difficulty_from_db(db_path, bid_row["climb_uuid"], bid_row["angle"])
-            displayed_grade = convert_difficulty_to_grade(difficulty, grades_dict, grade_type)
+        if aggregate:
+            entry = {
+                'uid': f"{bid_row['climb_uuid']}-{bid_row['angle']}",
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:547]: Let's stay consistent with `uuid`, confusing to drop the `u` here. If you want, it could also be `climb_angle_uuid` to be more descriptive. Maybe I'm missing something here, but couldn't Climbdex construct the climb + angle ID on that end so that boardlib didn't need to worry about this nuance?

File: src/boardlib/api/aurora.py [hunk 7]
```diff
@@ -367,31 +365,42 @@ def get_bids_logbook(board, token, user_id):
     sync_results = user_sync(board, token, user_id, tables=["bids"])
     return sync_results["PUT"].get("bids", [])
 
+def bids_logbook_entries(board, user_id, token, db_path=None, aggregate=False):
```
Comment (lemeryfertitta) [inline src/boardlib/api/aurora.py:368]: Can you explain what the goal is for the aggregate version? I thought the two changes needed for Climbdex were the UUID in the final output and the ability to use a token instead of username/password for auth. It seems like there is quite a bit more than this going on but it's a little hard to follow

**Review by lemeryfertitta — CHANGES_REQUESTED** [review]

File: src/boardlib/api/aurora.py (hunk 7 above)
Comment (lemeryfertitta) [reply src/boardlib/api/aurora.py:368]: Sounds good and no worries, happy to help if needed!
